### PR TITLE
[Attributor][FIX] Mark "may" accesses through call sites as such

### PR DESCRIPTION
**Pull request by jdoerfert**
Before, we kept the call site access kind (may/must) when we translated
the access. However, the pointer we access it through (by passing it to
the callee) might not be the underlying object. We have similar logic
when we add store and load accesses.

### CODE REVIEW

File: llvm/lib/Transforms/IPO/AttributorAttributes.cpp [hunk 1]
```diff
@@ -1893,9 +1898,10 @@ ChangeStatus AAPointerInfoFloating::updateImpl(Attributor &A) {
             DepClassTy::REQUIRED);
         if (!CSArgPI)
           return false;
-        Changed =
-            translateAndAddState(A, *CSArgPI, OffsetInfoMap[CurPtr], *CB) |
-            Changed;
+        bool IsMustAcc = (getUnderlyingObject(CurPtr) == &AssociatedValue);
+        Changed = translateAndAddState(A, *CSArgPI, OffsetInfoMap[CurPtr], *CB,
```
Comment (shiltian) [inline llvm/lib/Transforms/IPO/AttributorAttributes.cpp:1902]: Why not `Changed |= ...`?

File: llvm/test/Transforms/Attributor/IPConstantProp/2009-09-24-byval-ptr.ll [hunk 2]
```diff
@@ -139,12 +139,12 @@ define i32 @unions_v2() nounwind {
...
-; TUNIT-NEXT:    [[MYSTR_B4:%.*]] = getelementptr i8, ptr @mystr, i64 4
-; TUNIT-NEXT:    [[TMP1:%.*]] = load i32, ptr [[MYSTR_B4]], align 8
+; TUNIT-NEXT:    [[MYSTR_B41:%.*]] = getelementptr i8, ptr @mystr, i64 4
+; TUNIT-NEXT:    [[TMP1:%.*]] = load i32, ptr [[MYSTR_B41]], align 8
 ; TUNIT-NEXT:    call void @vfu1(i8 [[TMP0]], i32 [[TMP1]]) #[[ATTR2]]
 ; TUNIT-NEXT:    [[TMP2:%.*]] = load i8, ptr @mystr, align 8
-; TUNIT-NEXT:    [[MYSTR_B41:%.*]] = getelementptr i8, ptr @mystr, i64 4
-; TUNIT-NEXT:    [[TMP3:%.*]] = load i32, ptr [[MYSTR_B41]], align 8
+; TUNIT-NEXT:    [[MYSTR_B4:%.*]] = getelementptr i8, ptr @mystr, i64 4
+; TUNIT-NEXT:    [[TMP3:%.*]] = load i32, ptr [[MYSTR_B4]], align 8
```
Comment (shiltian) [inline llvm/test/Transforms/Attributor/IPConstantProp/2009-09-24-byval-ptr.ll:147]: pre-commit this?

File: llvm/test/Transforms/Attributor/ArgumentPromotion/byval.ll [hunk 3]
```diff
@@ -59,12 +59,12 @@ define i32 @main() nounwind  {
 ; TUNIT-NEXT:    store i32 1, ptr [[S]], align 32
 ; TUNIT-NEXT:    [[VAL4:%.*]] = getelementptr [[STRUCT_SS]], ptr [[S]], i32 0, i32 1
 ; TUNIT-NEXT:    [[TMP0:%.*]] = load i32, ptr [[S]], align 8
-; TUNIT-NEXT:    [[S_B41:%.*]] = getelementptr i8, ptr [[S]], i64 4
-; TUNIT-NEXT:    [[TMP1:%.*]] = load i64, ptr [[S_B41]], align 8
+; TUNIT-NEXT:    [[S_B4:%.*]] = getelementptr i8, ptr [[S]], i64 4
```
Comment (shiltian) [inline llvm/test/Transforms/Attributor/ArgumentPromotion/byval.ll:62]: pre-commit this?

File: llvm/test/Transforms/Attributor/IPConstantProp/return-constants.ll [hunk 4]
```diff
@@ -101,10 +101,10 @@ define i32 @caller2(i1 %Q) {
 ; CGSCC: Function Attrs: mustprogress nofree nosync nounwind willreturn memory(none)
 ; CGSCC-LABEL: define {{[^@]+}}@caller2
 ; CGSCC-SAME: (i1 noundef [[Q:%.*]]) #[[ATTR1]] {
-; CGSCC-NEXT:    [[X:%.*]] = call [[TMP0:%.*]] [[FOO:@[a-zA-Z0-9_$\"\\.-]*[a-zA-Z_$\"\\.-][a-zA-Z0-9_$\"\\.-]*]](i1 noundef [[Q]]) #[[ATTR2]]
+; CGSCC-NEXT:    [[X:%.*]] = call [[TMP0:%.*]] @[[FOO]](i1 noundef [[Q]]) #[[ATTR2]]
 ; CGSCC-NEXT:    [[A:%.*]] = extractvalue [[TMP0]] [[X]], 0
 ; CGSCC-NEXT:    [[B:%.*]] = extractvalue [[TMP0]] [[X]], 1
-; CGSCC-NEXT:    [[Y:%.*]] = call [[TMP0]] [[BAR:@[a-zA-Z0-9_$\"\\.-]*[a-zA-Z_$\"\\.-][a-zA-Z0-9_$\"\\.-]*]](i1 noundef [[Q]]) #[[ATTR2]]
+; CGSCC-NEXT:    [[Y:%.*]] = call [[TMP0]] @[[BAR:[a-zA-Z0-9_$\"\\.-]*[a-zA-Z_$\"\\.-][a-zA-Z0-9_$\"\\.-]*]](i1 noundef [[Q]]) #[[ATTR2]]
```
Comment (shiltian) [inline llvm/test/Transforms/Attributor/IPConstantProp/return-constants.ll:107]: pre-commit this?

File: llvm/test/Transforms/Attributor/cb_liveness_disabled.ll [hunk 5]
```diff
@@ -189,6 +189,7 @@ define i32 @test_ncheck2(i32 %0) #0 {
 
 attributes #0 = { noinline nounwind sspstrong uwtable}
 
+;.
```
Comment (shiltian) [inline llvm/test/Transforms/Attributor/cb_liveness_disabled.ll:192]: pre-commit this?

File: llvm/test/Transforms/Attributor/cb_liveness_enabled.ll [hunk 6]
```diff
@@ -192,6 +192,7 @@ define i32 @test_ncheck2(i32 %0) #0 {
 
 attributes #0 = { noinline nounwind sspstrong uwtable}
 
+;.
```
Comment (shiltian) [inline llvm/test/Transforms/Attributor/cb_liveness_enabled.ll:195]: unrelated change?

Comment (jdoerfert) [reply llvm/test/Transforms/Attributor/cb_liveness_enabled.ll:195]: It's in a standalone commit. I can push it first. Look at commit #3 only.

File: llvm/lib/Transforms/IPO/AttributorAttributes.cpp (hunk 1 above)
Comment (jdoerfert) [reply llvm/lib/Transforms/IPO/AttributorAttributes.cpp:1902]: Didn't change it from the old style, can do that.

**Review by shiltian — APPROVED** [review]
